pkg/channel/jetstream/dispatcher/pull_consumer.go

```diff
@@ -0,0 +1,329 @@
+	started     bool
+	closing     chan struct{}
+	closed      chan struct{}
+}
+
+func NewPullConsumer(ctx context.Context, consumer *nats.Subscription, subscription Subscription, dispatcher *kncloudevents.Dispatcher, reporter eventingchannels.StatsReporter, channelConfig *ChannelConfig) (*PullConsumer, error) {
+	consumerInfo, err := consumer.ConsumerInfo()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get consumer info: %w", err)
+	}
+
+	logger := logging.FromContext(ctx)
+	updatePullSubscriptionConfig(channelConfig, &subscription)
+
+	return &PullConsumer{
+		dispatcher:       dispatcher,
+		reporter:         reporter,
+		channelNamespace: channelConfig.Namespace,
+
+		natsConsumer:     consumer,
+		natsConsumerInfo: consumerInfo,
+
+		logger: logger,
+
+		sub: subscription,
+
+		closing: make(chan struct{}),
+		closed:  make(chan struct{}),
+	}, nil
+}
+
+func (c *PullConsumer) ConsumerType() ConsumerType {
+	return PullConsumerType
+}
+
+// Start begins the consumer and handles messages until Close is called. This method is blocking and
+// will return an error if the consumer fails prematurely. A nil error will be returned upon being
+// stopped by the Close method.
+func (c *PullConsumer) Start() error {
+	if err := c.checkStart(); err != nil {
+		return err
+	}
+	defer close(c.closed)
+
+	ctx := logging.WithLogger(context.Background(), c.logger)
+
+	var wg sync.WaitGroup
+
+	// wait for handlers to finish
+	defer wg.Wait()
+
+	// all messages are attached with this ctx, upon closing the consumer, we will cancel this ctx
+	// so that all pending messages are cancelled. This should cause any pending requests be
+	// cancelled (which also results in a nack) and the batch to be drained.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	for {
+		batch, err := c.natsConsumer.FetchBatch(FetchBatchSize, nats.MaxWait(c.sub.PullSubscription.FetchMaxWait))
+		if err != nil {
+			c.logger.Errorw("Failed to fetch messages", zap.Error(err), zap.String("consumer", c.sub.Name))
+		}
+
+		if err := c.consumeMessages(ctx, batch, &wg); err != nil {
+			if errors.Is(err, io.EOF) {
+				return nil
+			}
+
+			return err
+		}
+	}
+}
+
+// consumeMessages consumes messages from the batch and forwards them to the queue. The enqueued
+// message includes a context which will cancel after the AckWait duration of the consumer.
+//
+// This method returns once the MessageBatch has been consumed, or upon a call to Consumer.Close.
+// Returning as a result of Consumer.Close results in an io.EOF error.
+func (c *PullConsumer) consumeMessages(ctx context.Context, batch nats.MessageBatch, wg *sync.WaitGroup) error {
+	for {
+		select {
+		case natsMsg, ok := <-batch.Messages():
+			if !ok {
+				return nil
+			}
+
+			wg.Add(1)
+
+			go func() {
+				defer wg.Done()
+				ctx := logging.WithLogger(ctx, c.logger.With(zap.String("msg_id", natsMsg.Header.Get(nats.MsgIdHdr))))
+				msg := internal.NewMessage(ctx, natsMsg, c.natsConsumerInfo.Config.AckWait)
```

Comment:
do we need to call `msg.Finish()` somewhere?

Reply:
it is called inside message_dispatcher, https://github.com/knative-extensions/eventing-natss/blob/aabb79ea807ba93cd7ca53f469dfa1d459a30969/pkg/channel/jetstream/dispatcher/message_dispatcher.go#L137